src/uu/factor/src/factor.rs

```diff
@@ -3,302 +3,147 @@
+            // number.
+            show_warning!("{}: {}", num_str.maybe_quote(), e);
+            set_exit_code(1);
+            return Ok(());
         }
+    };
 
-        true
-    }
-}
-impl Eq for Decomposition {}
-
-#[derive(Clone, Debug, Eq, PartialEq)]
-pub struct Factors(RefCell<Decomposition>);
+    let (factorization, remaining) = if x > BigUint::from_u32(1).unwrap() {
+        num_prime::nt_funcs::factors(x.clone(), None)
+    } else {
+        (BTreeMap::new(), None)
+    };
 
-impl Factors {
-    pub fn one() -> Self {
-        Self(RefCell::new(Decomposition::one()))
+    if let Some(_remaining) = remaining {
+        return Err(USimpleError::new(
+            1,
+            "Factorization incomplete. Remainders exists.",
+        ));
     }
 
-    pub fn add(&mut self, prime: u64, exp: Exponent) {
-        debug_assert!(miller_rabin::is_prime(prime));
-        self.0.borrow_mut().add(prime, exp);
-    }
+    write_result(w, x, factorization, print_exponents).map_err_context(|| "write error".into())?;
 
-    pub fn push(&mut self, prime: u64) {
-        self.add(prime, 1);
-    }
-
-    #[cfg(test)]
-    fn product(&self) -> u64 {
-        self.0.borrow().product()
-    }
+    Ok(())
 }
 
-impl fmt::Display for Factors {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let v = &mut (self.0).borrow_mut().0;
-        v.sort_unstable();
-
-        let include_exponents = f.alternate();
-        for (p, exp) in v {
-            if include_exponents && *exp > 1 {
-                write!(f, " {p}^{exp}")?;
+fn write_result(
+    w: &mut io::BufWriter<impl Write>,
+    x: BigUint,
+    factorization: BTreeMap<BigUint, usize>,
+    print_exponents: bool,
+) -> io::Result<()> {
+    write!(w, "{x}:")?;
+    for (factor, n) in factorization {
+        if print_exponents {
+            if n > 1 {
+                write!(w, " {}^{}", factor, n)?;
             } else {
-                for _ in 0..*exp {
-                    write!(f, " {p}")?;
-                }
+                write!(w, " {}", factor)?;
             }
-        }
-
-        Ok(())
-    }
-}
-
-fn _factor<A: Arithmetic + miller_rabin::Basis>(num: u64, f: Factors) -> Factors {
-    use miller_rabin::Result::*;
-
-    // Shadow the name, so the recursion automatically goes from “Big” arithmetic to small.
-    let _factor = |n, f| {
-        if n < (1 << 32) {
-            _factor::<Montgomery<u32>>(n, f)
         } else {
-            _factor::<A>(n, f)
-        }
-    };
-
-    if num == 1 {
-        return f;
-    }
-
-    let n = A::new(num);
-    let divisor = match miller_rabin::test::<A>(n) {
-        Prime => {
-            #[cfg(feature = "coz")]
-            coz::progress!("factor found");
-            let mut r = f;
-            r.push(num);
-            return r;
+            for _i in 0..n {
+                write!(w, " {}", factor)?;
+            }
```

Comment:
can be simplify no? :)
```suggestion
write!(w, "{}", " ".repeat(n) + &factor.to_string())?;
```

Reply:
almost. see latest version. this works:

```rust
w.write_all(format!(" {}", factor).repeat(n).as_bytes())?;
```